File: examples/aws/ruby/00_auth_and_clients/client_with_sts_role.rb
require 'aws-sdk-bedrockruntime'
require 'aws-sdk-sts'
require 'dotenv'
require 'json'

# -----------------------------------------------------------------------------
# Helpers
# -----------------------------------------------------------------------------

def require_env(name)
  value = ENV[name]

  raise "Missing required environment variable: #{name}" if value.nil? || value.strip.empty?

  value
end

def disable_ssl_verification?
  ENV['DISABLE_SSL_VERIFICATION'].to_s.casecmp('true').zero?
end

# -----------------------------------------------------------------------------
# Lasso AWS Bedrock Converse example using STS AssumeRole.
#
# This example sends Bedrock Runtime requests to the Lasso proxy. Lasso then
# forwards the request to Amazon Bedrock.
#
# Auth flow:
#   1. The app uses local AWS credentials to call STS AssumeRole.
#   2. STS returns temporary role credentials.
#   3. The AWS SDK signs the Bedrock request with those temporary credentials.
#   4. A small SDK plugin adds the Lasso API key header.
#   5. Lasso receives the signed Bedrock request and forwards it to AWS.
#
# Required .env values:
#   AWS_REGION
#   AWS_ACCESS_KEY_ID
#   AWS_SECRET_ACCESS_KEY
#   AWS_ROLE_ARN
#   BEDROCK_TEXT_MODEL_ID
#   LASSO_PROXY_ENDPOINT
#   LASSO_X_API_KEY
#
# Optional .env values:
#   AWS_SESSION_TOKEN
#   AWS_ROLE_SESSION_NAME
#   AWS_EXTERNAL_ID
#   DISABLE_SSL_VERIFICATION=true
# -----------------------------------------------------------------------------

# -----------------------------------------------------------------------------
# 1. Load configuration
# -----------------------------------------------------------------------------

env_path = File.expand_path('../../../.env', __dir__)

raise "Missing .env file at #{env_path}" unless File.exist?(env_path)

Dotenv.load(env_path)

region = ENV.fetch('AWS_REGION', 'us-east-1')

source_access_key = require_env('AWS_ACCESS_KEY_ID')
source_secret_key = require_env('AWS_SECRET_ACCESS_KEY')
source_session_token = ENV['AWS_SESSION_TOKEN']

role_arn = require_env('AWS_ROLE_ARN')
role_session_name = ENV.fetch('AWS_ROLE_SESSION_NAME', 'lasso-ruby-bedrock-example')
external_id = ENV['AWS_EXTERNAL_ID']

model_id = require_env('BEDROCK_TEXT_MODEL_ID')
lasso_api_key = require_env('LASSO_X_API_KEY')
lasso_proxy_url = require_env('LASSO_PROXY_ENDPOINT').delete_suffix('/')

# The AWS SDK sends Bedrock Runtime requests to Lasso instead of AWS directly.
lasso_endpoint = "#{lasso_proxy_url}/v1/bedrock"

# -----------------------------------------------------------------------------
# 2. Assume the AWS role with STS
# -----------------------------------------------------------------------------

source_credentials = Aws::Credentials.new(
  source_access_key,
  source_secret_key,
  source_session_token
)

sts_client = Aws::STS::Client.new(
  region: region,
  credentials: source_credentials,
  ssl_verify_peer: !disable_ssl_verification?
)

assume_role_params = {
  role_arn: role_arn,
  role_session_name: role_session_name
}

# Use ExternalId when the target role trust policy requires it.
assume_role_params[:external_id] = external_id if external_id && !external_id.strip.empty?

assumed_role = sts_client.assume_role(assume_role_params)

role_credentials = Aws::Credentials.new(
  assumed_role.credentials.access_key_id,
  assumed_role.credentials.secret_access_key,
  assumed_role.credentials.session_token
)

# -----------------------------------------------------------------------------
# 3. Add the Lasso API key to each Bedrock request
# -----------------------------------------------------------------------------

# The AWS SDK signs the request. This plugin only adds the Lasso API key header.
class LassoHeaderPlugin < Seahorse::Client::Plugin
  option(:lasso_api_key)

  class Handler < Seahorse::Client::Handler
    def call(context)
      context.http_request.headers['lasso-x-api-key'] =
        context.config.lasso_api_key

      @handler.call(context)
    end
  end

  handler(Handler, step: :build)
end

Aws::BedrockRuntime::Client.add_plugin(LassoHeaderPlugin)

# -----------------------------------------------------------------------------
# 4. Create the Bedrock Runtime client through Lasso
# -----------------------------------------------------------------------------

client = Aws::BedrockRuntime::Client.new(
  region: region,
  endpoint: lasso_endpoint,

  # These are the temporary credentials returned by STS AssumeRole.
  credentials: role_credentials,

  lasso_api_key: lasso_api_key,

  # Local testing only.
  # Do not disable certificate validation in production.
  ssl_verify_peer: !disable_ssl_verification?
)

# -----------------------------------------------------------------------------
# 5. Send a Bedrock Converse request
# -----------------------------------------------------------------------------

response = client.converse(
  model_id: model_id,
  messages: [
    {
      role: 'user',
      content: [
        {
          text: 'In a single word, what is the answer to the great question of life, the universe and everything?'
        }
      ]
    }
  ],
  inference_config: {
    max_tokens: 512
  }
)

# -----------------------------------------------------------------------------
# 6. Print response
# -----------------------------------------------------------------------------

puts "HTTP status code: #{response.context.http_response.status_code}"

answer = response.to_h.dig(:output, :message, :content, 0, :text)

puts
puts 'Answer:'
puts answer

File: examples/aws/python/00_auth_and_clients/client_with_sts_role.py
"""
Simple AWS Bedrock Converse example through the Lasso proxy using an STS assumed role.

Required environment variables:
  LASSO_PROXY_ENDPOINT
  LASSO_X_API_KEY
  AWS_REGION
  AWS_ROLE_ARN
  AWS_ROLE_SESSION_NAME  # optional, defaults to lasso-bedrock-example
  BEDROCK_MODEL_ID

AWS credentials must already be available through one of the standard boto3
credential sources, such as:
  - AWS profile
  - environment variables
  - EC2 instance role
  - ECS task role
  - EKS pod identity / IRSA
"""

import json
import logging
import os

import boto3
import botocore.exceptions
from botocore.config import Config
from dotenv import load_dotenv


logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

load_dotenv()


# 1. Load configuration
lasso_proxy_url = os.getenv("LASSO_PROXY_ENDPOINT")
lasso_api_key = os.getenv("LASSO_X_API_KEY")

region = os.getenv("AWS_REGION", "us-east-1")
role_arn = os.getenv("AWS_ROLE_ARN")
role_session_name = os.getenv("AWS_ROLE_SESSION_NAME", "lasso-bedrock-example")

model_id = os.getenv("BEDROCK_TEXT_MODEL_ID")


# 2. Assume the AWS role with STS
sts_client = boto3.client("sts", region_name=region)

assumed_role = sts_client.assume_role(
    RoleArn=role_arn,
    RoleSessionName=role_session_name,
)

credentials = assumed_role["Credentials"]


# 3. Create Bedrock client with the temporary role credentials and point it to Lasso
bedrock_client = boto3.client(
    "bedrock-runtime",
    region_name=region,
    endpoint_url=f"{lasso_proxy_url}/v1/bedrock",
    aws_access_key_id=credentials["AccessKeyId"],
    aws_secret_access_key=credentials["SecretAccessKey"],
    aws_session_token=credentials["SessionToken"],
    config=Config(
        retries={"max_attempts": 1, "mode": "standard"},
        connect_timeout=5,
        read_timeout=12,
        user_agent_extra="lasso-proxy-example/1.0",
    ),
    verify=False
)


# 4. Add the Lasso API key header to every Bedrock request
def add_lasso_header(request, **kwargs):
    request.headers.add_header("lasso-x-api-key", lasso_api_key)


bedrock_client.meta.events.register_first(
    "before-sign.bedrock-runtime.*",
    add_lasso_header,
)


# 5. Call Bedrock Converse through Lasso
try:
    response = bedrock_client.converse(
        modelId=model_id,
        messages=[
            {
                "role": "user",
                "content": [
                    {
                        "text": "What is the ultimate answer to life, the universe, and everything?"
                    }
                ],
            }
        ],
        inferenceConfig={
            "maxTokens": 512,
        },
    )

    print(json.dumps(response, indent=2, default=str))

except botocore.exceptions.ClientError as error:
    logger.exception("Bedrock request failed: %s", error)